pkg/http/api.go

```diff
@@ -2020,6 +2026,232 @@ func (a *api) onPublish(reqCtx *fasthttp.RequestCtx) {
+	// Populate W3C tracestate to cloudevent envelope
+	traceState := diag.TraceStateToW3CString(span.SpanContext())
+
+	incomingEntries := make([]bulkPublishMessageEntry, 0)
+	err := json.Unmarshal(body, &incomingEntries)
+	if err != nil {
+		msg := NewErrorResponse("ERR_PUBSUB_EVENTS_SER",
+			fmt.Sprintf(messages.ErrPubsubUnmarshal, topic, pubsubName, err.Error()))
+		respond(reqCtx, withError(fasthttp.StatusBadRequest, msg))
+		log.Debug(msg)
+
+		return
+	}
+	entries := make([]pubsub.BulkMessageEntry, len(incomingEntries))
+
+	for i, entry := range incomingEntries {
+		log.Debugf("Incoming event:  %v\n", entry)
+		var dBytes []byte
+		if contribContentType.IsStringContentType(entry.DataContentType) || contribContentType.IsBinaryContentType(entry.DataContentType) {
+			switch v := entry.Event.(type) {
+			case string:
+				dBytes = []byte(v)
+			case []byte:
+				dBytes = v
+			default:
+				msg := NewErrorResponse("ERR_PUBSUB_EVENTS_SER",
+					fmt.Sprintf(messages.ErrPubsubMarshal, topic, pubsubName, "error: mismatch between dataContentType and event"))
+				respond(reqCtx, withError(fasthttp.StatusBadRequest, msg))
+				log.Debug(msg)
+
+				return
+			}
+		} else {
+			dBytes, err = json.Marshal(entry.Event)
+			if err != nil {
+				msg := NewErrorResponse("ERR_PUBSUB_EVENTS_SER",
+					fmt.Sprintf(messages.ErrPubsubMarshal, topic, pubsubName, err.Error()))
+				respond(reqCtx, withError(fasthttp.StatusBadRequest, msg))
+				log.Debug(msg)
+
+				return
+			}
+		}
+		entries[i] = pubsub.BulkMessageEntry{
+			Event:       dBytes,
+			ContentType: entry.DataContentType,
+		}
+		if entry.Metadata != nil {
+			entries[i].Metadata = entry.Metadata
+		}
+		if entry.EntryID == "" {
+			msg := NewErrorResponse("ERR_PUBSUB_EVENTS_SER",
+				fmt.Sprintf(messages.ErrPubsubMarshal, topic, pubsubName, "error: entryID is not present for entry"))
+			respond(reqCtx, withError(fasthttp.StatusBadRequest, msg))
+			log.Debug(msg)
+
+			return
+		}
+		entries[i].EntryID = entry.EntryID
+	}
+
+	spanMap := map[int]trace.Span{}
+	// closeChildSpans method is called on every respond() call in all return paths in the following block of code.
+	closeChildSpans := func(ctx *fasthttp.RequestCtx) {
+		for _, span := range spanMap {
+			diag.UpdateSpanStatusFromHTTPStatus(span, ctx.Response.StatusCode())
+			span.End()
+		}
+	}
+	if !rawPayload {
+		for i := range entries {
+			// For multiple events in a single bulk call traceParent is different for each event.
+			childSpan := diag.StartProducerSpanChildFromParent(reqCtx, span)
+			// Populate W3C traceparent to cloudevent envelope
+			corID := diag.SpanContextToW3CString(childSpan.SpanContext())
+			spanMap[i] = childSpan
+
+			var envelope map[string]interface{}
+
+			envelope, err = runtimePubsub.NewCloudEvent(&runtimePubsub.CloudEvent{
+				ID:              a.id,
+				Topic:           topic,
+				DataContentType: entries[i].ContentType,
+				Data:            entries[i].Event,
+				TraceID:         corID,
```

Comment:
@yaron2 @artursouza here a new span and trace parent header is added for every event inside that bulk publish request.